File: src/j/implementation/microSet/InnerSet_3.java
package j.implementation.microSet;

import java.util.Iterator;

public class InnerSet_3 <T> extends AbstractInnerSet<T>
{
        protected T element_1;
        protected T element_2;
        protected T element_3;
        
    public InnerSet_3(T p_element_1, T p_element_2, T p_element_3)
    {
                element_1 = p_element_1;
                element_2 = p_element_2;
                element_3 = p_element_3;
            }

    public InnerSet_3(InnerSet_3<T> i)
    {
            element_1 = i.element_1;
            element_2 = i.element_2;
            element_3 = i.element_3;
        }

    public InnerSet_3<T> copy() {
        return this;
    }

    public InnerSet_3(InnerSet_2<T> i, T el)
    {
            element_1 = i.element_1;            element_2 = i.element_2;                    element_3 = el;
    }
    
    @Override
    public InnerSet<T> addElement(T p_element)
    {
        if(element_1.equals(p_element) || element_2.equals(p_element) || element_3.equals(p_element))
            return this;
        else
            return new InnerSet_4<>(this, p_element);    }

    @Override
    public InnerSet<T> addUnChecked(T p_element)
    {
        return new InnerSet_4<>(this, p_element);    }

    @Override
    public InnerSet<T> removeElement(Object p_element)
    {
                if(element_1.equals(p_element))
            return  new InnerSet_2<>(element_2, element_3);
         else                 if(element_2.equals(p_element))
            return  new InnerSet_2<>(element_1, element_3);
         else                 if(element_3.equals(p_element))
            return  new InnerSet_2<>(element_1, element_2);
                        else
            return this;
    }

    @Override
    public boolean containsElement(Object p_element)
    {
        return element_1.equals(p_element) || element_2.equals(p_element) || element_3.equals(p_element);
    }


    public T getElement(int index)
    {switch(index) {
                case 0:
            return element_1;
                case 1:
            return element_2;
                case 2:
            return element_3;
                default:
            return null;
            }
    }


    public InnerSet<T> addAllElements(InnerSet<T> innerSet) {
        // Reverse the add if the innerSet we add is smaller than this
        if (innerSet.getSize() < this.getSize()) {
            return innerSet.addAllElements(this);
        }
        return innerSet.addElement(element_1).addElement(element_2).addElement(element_3);
    }


    @Override
    public InnerSet<T> removeAllElements(InnerSet<T> innerSet) {
        return innerSet.removeElement(element_1).removeElement(element_2).removeElement(element_3);
    }

    @Override
    public InnerSet<T> retainAllElements(InnerSet<T> innerSet) {
        InnerSet<T> i = this;
                if (!innerSet.containsElement(element_1)) i = i.removeElement(element_1);
                if (!innerSet.containsElement(element_2)) i = i.removeElement(element_2);
                if (!innerSet.containsElement(element_3)) i = i.removeElement(element_3);
                return i;
    }

    @Override
    public boolean containsAllElements(InnerSet<T> innerSet) {
        return  innerSet.containsElement(element_1) && innerSet.containsElement(element_2) && innerSet.containsElement(element_3);
    }

    public InnerSet<T> addAllAndPropagate(InnerSet<T> innerSet, MicroSet<T> microSetToPropagate) {
        return innerSet.addAllAndPropagateReverse(this, microSetToPropagate);
    }


    public InnerSet<T> addAllAndPropagateReverse(InnerSet<T> innerSet, MicroSet<T> microSetToPropagate) {
        return innerSet.addAndPropagate(element_1, microSetToPropagate).addAndPropagate(element_2, microSetToPropagate).addAndPropagate(element_3, microSetToPropagate);
    }

    public InnerSet<T> addAndPropagate(T p_element, MicroSet<T> microSetToPropagate) {
        if(element_1.equals(p_element) || element_2.equals(p_element) || element_3.equals(p_element))
            return this;
        else {
            microSetToPropagate.add(p_element);
            return new InnerSet_4<>(this, p_element);        }
    }

    @Override
    public String toString()
    {
        return "{" + element_1 + ", " + element_2 + ", " + element_3 + " }";
    }
    
    @Override
    public int getSize()
    {
        return 3;
    }

    @Override
    public Iterator<T> iterator() {
        return new InnerSetIterator<>(this);
    }

    @Override
    public InnerSet<T> clear(boolean unused) {
        return (InnerSet<T>) InnerSet_0.singleton;
    }
}

File: src/cpp/implementation/save/save/inner_set_3.cpp
#include "inner_set_0.hpp"
#include "inner_set_3.hpp"

#include "inner_array_set.hpp"

#include <cstdlib>
#include <ctime>
#include <iostream>

template<typename T, int p_to_up>
inner_set_3<T, p_to_up>::inner_set_3()
{

}

template<typename T, int p_to_up>
inner_set_3<T, p_to_up>::inner_set_3(T p_element_1, T p_element_2, T p_element_3)
{
            a_values.element_1 = p_element_1;
            a_values.element_2 = p_element_2;
            a_values.element_3 = p_element_3;
    }

template<typename T, int p_to_up>
inner_set_3<T, p_to_up>::inner_set_3(inner_set_2<T, p_to_up>* p_inner_set, T p_element)
{
            a_values.element_1 = p_inner_set->a_values.element_1;
            a_values.element_2 = p_inner_set->a_values.element_2;

            a_values.element_3 = p_element;

            delete p_inner_set;
}

template<typename T, int p_to_up>
inner_set_3<T, p_to_up>::~inner_set_3()
{

}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_3<T, p_to_up>::add_element(T p_element)
{
  if(!contains_element(p_element))
    if(this->a_to_up == 3)
      return new inner_array_set<T, p_to_up>(this, p_element);
    else
      return new inner_array_set<T, p_to_up>(this, p_element); //TODO 4
  else
    return this;
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_3<T, p_to_up>::add_all_elements(inner_set<T, p_to_up>* p_inner_set)
{
  return p_inner_set->add_element(a_values.element_1)->add_element(a_values.element_2)->add_element(a_values.element_3);
}

template<typename T, int p_to_up>
bool inner_set_3<T, p_to_up>::contains_element(T p_element)
{
    return  a_values.element_1 == p_element ||  a_values.element_2 == p_element ||  a_values.element_3 == p_element;
}

template<typename T, int p_to_up>
bool inner_set_3<T, p_to_up>::contains_all_elements(inner_set<T, p_to_up>* p_inner_set)
{
  return p_inner_set->contains_element(a_values.element_1) &&
    p_inner_set->contains_element(a_values.element_2);
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>*  inner_set_3<T, p_to_up>::copy()
{
	return this;
}

template<typename T, int p_to_up>
T inner_set_3<T, p_to_up>::get_element(int p_index)
{
    return reinterpret_cast<T*>(&a_values)[p_index];
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_3<T, p_to_up>::remove_element(T p_element)
{
  T t_value_1 = {};
  T t_value_2 = {};

  if(a_values.element_1 == p_element)
  {
    t_value_1 = a_values.element_2;
    t_value_2 = a_values.element_3;

    delete this;

    return new inner_set_2<T, p_to_up>(t_value_1, t_value_2);
  }
  else if(a_values.element_2 == p_element)
  {
    t_value_1 = a_values.element_1;
    t_value_2 = a_values.element_3;

    delete this;

    return new inner_set_2<T, p_to_up>(t_value_1, t_value_2);
  }
  else if(a_values.element_3 == p_element)
  {
    t_value_1 = a_values.element_1;
    t_value_2 = a_values.element_2;

    delete this;

    return new inner_set_2<T, p_to_up>(t_value_1, t_value_2);
  }
  else
    return this;
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_3<T, p_to_up>::remove_all_elements(inner_set<T, p_to_up>* p_inner_set)
{
  return p_inner_set->remove_element(a_values.element_1)->remove_element(a_values.element_2)->remove_element(a_values.element_3);
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_3<T, p_to_up>::retain_all_elements(inner_set<T, p_to_up>* p_inner_set)
{
  inner_set<T, p_to_up>* r_inner_set = inner_set_0<T, p_to_up>::EMPTY;

  if(p_inner_set->contains_element(a_values.element_1))
    r_inner_set = r_inner_set->add_element(a_values.element_1);

  if(p_inner_set->contains_element(a_values.element_2))
    r_inner_set = r_inner_set->add_element(a_values.element_2);

  if(p_inner_set->contains_element(a_values.element_3))
    r_inner_set = r_inner_set->add_element(a_values.element_3);

    if(this->get_size() != 0)
      delete this;

  return r_inner_set;
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_3<T, p_to_up>::add_and_propagate(T p_element, micro_set<T, p_to_up>* p_micro_set)
{
  if(a_values.element_1 == p_element ||
    a_values.element_2 == p_element ||
    a_values.element_3 == p_element)
    return this;
  else
  {
	   p_micro_set->insert(p_element);

	    return new inner_array_set<T, p_to_up>(this, p_element);
  }
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_3<T, p_to_up>::add_all_and_propagate(inner_set<T, p_to_up>* p_inner_set, micro_set<T, p_to_up>* p_micro_set)
{
  return p_inner_set->add_all_and_propagate_reverse(this, p_micro_set);
}

template<typename T, int p_to_up>
inner_set<T, p_to_up>* inner_set_3<T, p_to_up>::add_all_and_propagate_reverse(inner_set<T, p_to_up>* p_inner_set, micro_set<T, p_to_up>* p_micro_set)
{
  return p_inner_set->add_and_propagate(a_values.element_1, p_micro_set)->add_and_propagate(a_values.element_2, p_micro_set)->add_and_propagate(a_values.element_3, p_micro_set);
}

template<typename T, int p_to_up>
int inner_set_3<T, p_to_up>::get_size()
{
    return 3;
}
